core/core.go

```diff
@@ -346,7 +347,35 @@ func (n *IpfsNode) OnlineMode() bool {
 }
 
 func (n *IpfsNode) Resolve(fpath string) (*merkledag.Node, error) {
-	return n.Resolver.ResolvePath(path.Path(fpath))
+	// By default we assume the path is <ipfs>
+	pathToResolve := path.Path(fpath)
+
+	// If it looks like an <ipns>, we re-write the path
+	// NOTE logic is partially duplicated, likely target for injection
+	if n.Namesys.CanResolve(fpath) {
+		key, err := n.Namesys.Resolve(n.Context(), fpath)
+
+		if err != nil {
+			return nil, err
+		} else {
+			// We found something, let's rewrite the path as
+			// ipfs/<ipns-key>/more/folders
+			parts := path.Path(fpath).Segments()
+			if parts[0] == "ipns" {
+				if len(parts) == 1 {
+					parts = parts[1:]
+				} else {
+					parts = parts[2:]
+				}
+			}
+			parts = append([]string{"ipfs", key.String()}, parts...)
+
+			pathToResolve = path.Path(strings.Join(parts, "/"))
+		}
+	}
+
+	// At this point, the value of fPath should be resolvable as a merkledag
+	return n.Resolver.ResolvePath(pathToResolve)
```

Comment:
I think we want to take this moment to:
- make this a function (like the core/* functions) that takes a node
- make some simplifying assumptions from the path:
  - require /ipns/... for ipns paths.
  - assume ipfs otherwise.
- expand our `path` package with utils

```go


func Resolve(n *IpfsNode, p path.Path) (*merkledag.Node, error) {
	strpath := string(p)

	// for now, we only try to resolve ipns paths if 
	// they begin with "/ipns/". Otherwise, ambiguity
	// emerges when resolving just a <hash>. Is it meant
	// to be an ipfs or an ipns resolution?

	if strings.HasPrefix(strpath, "/ipns/") {
		// if it's an ipns path, try to resolve it. 
		// if we can't, we can give that error back to the user.
		key, err := n.Namesys.Resolve(n.Context(), p)
		if err != nil {
			return nil, err
		}

		nr := ipath.FromSegments("ipfs", key.Pretty())
		p = path.RebasePath(p, nr)
	}

	// ok, we have an ipfs path now (or what we'll treat as one)
	return n.Resolver.ResolvePath(p)
}

// in our path/ pkg:
func RebasePath(p path.Path, newRoot path.Path) path.Path {
	parts := p.Segments()

	// skip over first /ipfs or /ipns
	if len(parts) > 0 && IsProtoComponent(parts[0]) {
		parts = parts[1:]
	}

	// skip the root we're rebasing
	if len(parts) > 0 {
		parts = parts[1:]
	}

	rparts := newRoot.Segments()
	return FromSegments(append(rparts, parts...)...)
}

func IsProtoComponent(s string) bool {
	return s == "ipfs" || s == "ipns"
}

func FromSegments(seg ...string) path.Path {
	return "/" + path.Join(seg...)
}
```

warning: i haven't compiled/tested this. may have a bug.